rhel/rhel.go

```diff
@@ -0,0 +1,329 @@
+		return nil, hint, driver.Unchanged
+	case http.StatusOK:
+		// break
+	default:
+		res.Body.Close()
+		return nil, hint, fmt.Errorf("rhel: fetcher got unexpected HTTP response: %d (%s)", res.StatusCode, res.Status)
+	}
+
+	if t := res.Header.Get("Last-Modified"); t != "" {
+		hint = driver.Fingerprint(t)
+	}
+	return res.Body, hint, nil
+}
+
+// Parse satisifies the driver.Updater interface.
+func (u *Updater) Parse(r io.ReadCloser) ([]*claircore.Vulnerability, error) {
+	return u.ParseContext(context.Background(), r)
+}
+
+// DistMap is a helper to prevent making thousands of Distribution objects.
+type distMap map[int]*claircore.Distribution
+
+func newDistMap() distMap {
+	return distMap(make(map[int]*claircore.Distribution))
+}
+func (m distMap) Version(v int) *claircore.Distribution {
+	d, ok := m[v]
+	if !ok {
+		d = &claircore.Distribution{}
+		d.VersionID = strconv.Itoa(v)
+		d.Name = "Red Hat Enterprise Linux"
+		d.Version = fmt.Sprintf("%s %s", d.Name, d.VersionID)
+		m[v] = d
+	}
+	return d
+}
+
+// ParseContext is like Parse, but with context.
+func (u *Updater) ParseContext(ctx context.Context, r io.ReadCloser) ([]*claircore.Vulnerability, error) {
+	defer r.Close()
+	root := oval.Root{}
+	if err := xml.NewDecoder(r).Decode(&root); err != nil {
+		return nil, fmt.Errorf("rhel: unable to decode OVAL document: %w", err)
+	}
+
+	distMap := newDistMap()
+	vs := make([]*claircore.Vulnerability, 0, len(root.Definitions.Definitions))
+
+	for _, def := range root.Definitions.Definitions {
+		links := []string{}
+		for _, ref := range def.References {
+			links = append(links, ref.RefURL)
+		}
+		for _, ref := range def.Advisory.Refs {
+			links = append(links, ref.URL)
+		}
+		for _, bug := range def.Advisory.Bugs {
+			links = append(links, bug.URL)
+		}
+		// It's likely that we'll have multiple vulnerabilites spawned from once
+		// CVE/definition, so this constructs new records on demand.
+		mkVuln := func() *claircore.Vulnerability {
+			return &claircore.Vulnerability{
+				Updater:     u.Name(),
+				Name:        def.References[0].RefID,
+				Description: def.Description,
+				Severity:    def.Advisory.Severity,
+				Links:       strings.Join(links, " "),
+				Package:     &claircore.Package{},
+			}
+		}
+
+		crit, err := walk(&def.Criteria)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, crit := range crit {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
+			v := mkVuln()
+			var rt *oval.RPMInfoTest
+			for _, c := range crit {
+				k, i, err := root.Tests.Lookup(c.TestRef)
+				if err != nil {
+					return nil, fmt.Errorf("rhel: dangling ref: %w", err)
+				}
+				switch k {
+				case "rpminfo_test": // This one
+					rt = &root.Tests.RPMInfoTests[i]
+				default:
+					continue
+				}
+				if len(rt.ObjectRefs) != 1 {
+					return nil, fmt.Errorf("unsure how to handle multple object references in a test")
+				}
+
+				objRef := rt.ObjectRefs[0].ObjectRef
```

Comment:
Would be nice to make a set of functions which are responsible for pulling out, RMPInfoTests, Obj, and StateRefs, so that this parse method really just calls methods which do the OVAL stuff, spawns a Vulnerability, and attaches data to it.